### PR TITLE
[fix](nereids) Fix the bugs of data distribution calculation on OlapScan

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.54 seconds
 load time: 478 seconds
 storage size: 17122573549 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230107094038_clickbench_pr_75328.html

File: regression-test/suites/nereids_syntax_p0/scan_partition.groovy [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    sql """INSERT INTO test_part VALUES (1, 1, '1');"""
+    sql """INSERT INTO test_part VALUES (1, 1, '1');"""
+    sql """INSERT INTO test_part VALUES (1, 5, '5');"""
+    sql """INSERT INTO test_part VALUES (1, 5, '5');"""
+    sql """INSERT INTO test_part VALUES (2, 5, '5');"""
+    sql """INSERT INTO test_part VALUES (3, 5, '1');"""
+    sql """INSERT INTO test_part VALUES (4, 5, '5');"""
+    sql """INSERT INTO test_part VALUES (7, 5, '9');"""
+    sql "SET enable_nereids_planner=true"
+    sql "SET enable_fallback_to_original_planner=false"
```
Comment (morrySnow) [inline regression-test/suites/nereids_syntax_p0/scan_partition.groovy:44]: should set `parallel_fragment_exec_instance_num` to `8`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java [hunk 2]
```diff
@@ -59,7 +63,17 @@ public Rule build() {
...
+        ColocateTableIndex colocateTableIndex = Env.getCurrentColocateIndex();
+        if ((colocateTableIndex.isColocateTable(olapTable.getId())
+                && !colocateTableIndex.isGroupUnstable(colocateTableIndex.getGroup(olapTable.getId())))
+                || olapTable.getPartitionInfo().getType() == PartitionType.UNPARTITIONED
+                || olapTable.getPartitions().size() == 1) {
+            if (!(distributionInfo instanceof HashDistributionInfo)) {
+                return DistributionSpecAny.INSTANCE;
+            }
+        }
         if (distributionInfo instanceof HashDistributionInfo) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java:77]: we should return `DistributionSpecAny.INSTANCE` even if table's `defaultDistributionInfo` is `HashDistributionInfo`, when we need to scan more than one partition and table is not colocate table or not in a stable group.

Comment (Kikyou1997) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalOlapScanToPhysicalOlapScan.java:77]: done

File: regression-test/suites/nereids_syntax_p0/scan_partition.groovy (hunk 1 above)
Comment (Kikyou1997) [reply regression-test/suites/nereids_syntax_p0/scan_partition.groovy:44]: done

**Review by morrySnow — APPROVED** [review]

**Pull request merged**